examples/amm/lib.rs

```diff
@@ -0,0 +1,725 @@
+//!
+//! ## Error Handling
+//!
+//! Any function that modifies the state returns a `Result` type and does not changes the
+//! state if the `Error` occurs. The errors are defined as an `enum` type. Any other error
+//! or invariant violation triggers a panic and therefore rolls back the transaction.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod amm {
+    use ink::storage::Mapping;
+    use scale::{Decode, Encode};
+
+    /// Amm
+    /// This contract maintains all three tokens: token1, token2, and lp_tokens
+    /// total_lp_tokens - the aggregate total of all LP tokens created / burned
+    /// total_token1 - the total liquidity for token1
+    /// total_token2 - the total liquidity for token2
+    /// total1_balances - a ledger containing all user balances for token1
+    /// total2_balances - a ledger containing all user balances for token2
+    /// lp_token_balances - a ledger containing all user balances for the LP token
+    /// fee - the LP cut for all swaps made in the pool
+    #[derive(Default)]
+    #[ink(storage)]
+    pub struct Amm {
+        total_lp_tokens: LPTokens,
+        total_token1: Token1,
+        total_token2: Token2,
+        token1_balances: Mapping<AccountId, Token1>,
+        token2_balances: Mapping<AccountId, Token2>,
+        lp_token_balances: Mapping<AccountId, LPTokens>,
+        fee: Balance,
+        precision: u128,
+    }
+
+    /// Here we create type aliases for readibility
+    type Token1 = Balance;
+    type Token2 = Balance;
+    type LPTokens = Balance;
+
+    /// Errors are returned for the following conditions:
+    /// ZeroLiquidity - the caller does not have any LP tokens
+    /// ZeroAmount - the caller has a zero balance for the token in question
+    /// InsufficientAmount - the caller sends a transaction exceeding their balance
+    /// NonEquivalentAmount - an LP deposits two tokens with unequal value
+    /// ThresholdNotReached - the calculated LP token amount for a contribution is 0
+    /// InvalidLPAmount - a caller withdraws more than the value of their LP balance
+    /// InsufficientLiquidity - swap amount exceeds what is available in the pool
+    /// SlippageExceeded - the max or min slippage exceeds the threshold set by the caller
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        InsufficientAmount,
+        InsufficientLiquidity,
+        InvalidLPAmount,
+        NonEquivalentAmount,
+        SlippageExceeded,
+        ThresholdNotReached,
+        ZeroAmount,
+        ZeroLiquidity,
+    }
+
+    /// We create an ergonomic helper so we can do things like Result<Balance> instead of
+    /// Result<Balance, Error>
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    /// Struct to hold Account balances
+    #[derive(Default, Copy, PartialEq, Eq, Debug, Clone, Decode, Encode)]
```

Comment:
We should be using the [`#[ink(storage_item)`](https://paritytech.github.io/ink/ink/attr.storage_item.html) macro anything that goes into storage instead of manually adding all these derives